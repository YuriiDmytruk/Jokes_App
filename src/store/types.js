export const ADD_JOKES = 'ADD_JOKES';
export const GET_JOKES = 'GET_JOKES';
