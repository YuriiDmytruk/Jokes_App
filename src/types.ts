export type Joke = {
  joke: string;
  category: string;
  id: number;
}

export type JokesState = {
  jokes: Joke[],
};

export type JokesActionAdd = {
  type: string,
  jokes?: Joke[],
};

export type JokeDelete = {
  type: string,
  id: number,
}

export type JokesActionFetch = {
  type: string,
  props: {
      jokesLastID: number,
      jokesAmount: number,
    }
};

export type JokesControlProps = {
  jokesLength: number;
  all: string;
  categories: string[];
  jokesLastID: number;
  setPage: (page: number) => void;
  setFilter: (filter: string) => void;
};

export type JokesPaginationProps = {
  setPage: (page: number) => void;
  jokesLength: number;
  JOKES_ON_PAGE: number;
  page: number;
};
