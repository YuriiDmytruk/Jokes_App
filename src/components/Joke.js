import React from 'react';
import { Link } from 'react-router-dom';
import { Button, Card, Col, Container } from 'react-bootstrap';

const cardSize = {
  width: '17rem',
  height: '10rem',
  margin: '10px auto',
};
const cardTextSize = { height: '3rem' };
const CHARS_IN_JOKE = 50;
const DOTS = '...';

export default function Joke(props) {
  return (
    <Col>
      <Container>
        <Card style={cardSize}>
          <Card.Body>
            <Card.Title tag="h5">{props.joke.category}</Card.Title>
            <Card.Text style={cardTextSize}>
              {props.joke.joke.slice(0, CHARS_IN_JOKE) + DOTS}
            </Card.Text>
            <Link to={`/joke/${props.joke.id}`}>
              <Button color="primary">Check</Button>
            </Link>
          </Card.Body>
        </Card>
      </Container>
    </Col>
  );
}
