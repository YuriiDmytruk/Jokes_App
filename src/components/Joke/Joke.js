import React from "react";

import {
  Wrapper,
  CategoryPlace,
  CategoryText,
  ButtonPlace,
  Button,
  JokeText,
} from "./styledComponents";

export default function Joke(props) {
  const checkClickHendler = () => {
    console.log(props.joke.id);
  };

  return (
    <Wrapper key={props.joke.id}>
      <CategoryPlace>
        <CategoryText>{props.joke.category}</CategoryText>
      </CategoryPlace>
      <JokeText>
        {props.showAll ? props.joke.joke : props.joke.joke.slice(0, 50) + "..."}
      </JokeText>
      <ButtonPlace>
        <Button onClick={checkClickHendler}>Check</Button>
      </ButtonPlace>
    </Wrapper>
  );
}

/*

*/
