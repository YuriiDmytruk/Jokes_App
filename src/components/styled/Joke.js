import styled from 'styled-components';

export const CardSize = styled.div`
  width: 17rem;
  height: 10rem;
  margin: 10px auto;
`;

export const CardTextSize = styled.div`
  height: 3rem;
`;
