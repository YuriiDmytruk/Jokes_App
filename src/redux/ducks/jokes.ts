import { Joke, JokesState, JokesActionFetch, JokesActionAdd, JokeDelete } from '../../types'
import { put, takeEvery, call } from 'redux-saga/effects';
import { SagaIterator } from 'redux-saga';

import { fetchJokes as fetchJokesFromApi } from '../../api';

export const ADD_JOKES = 'ADD_JOKES';
export const FETCH_JOKES = 'FETCH_JOKES';
export const SET_JOKES = 'SET_JOKES';
export const DELETE_JOKE = 'DELETE_JOKE'

export const JOKES = 'JOKES'

const defaultState: JokesState = {
  jokes: JSON.parse(localStorage.getItem(JOKES)!)
};

export const jokesReducer = (
  state: JokesState = defaultState,
  action: any
): JokesState => {
  switch (action.type) {
    case ADD_JOKES:
      return { ...state, jokes: state.jokes.concat(action.jokes || []) };
    case SET_JOKES:
      return { ...state, jokes: action.jokes };
    case DELETE_JOKE:
      return { ...state, jokes: state.jokes.filter(joke => joke.id !== action.id) };
    default:
      return state;
  }
};

export const deleteJoke = (id: number): JokeDelete => {
  return {type: DELETE_JOKE, id: id}
}

export const setJokes = (jokes: Joke[]): JokesActionAdd => {
  return { type: SET_JOKES, jokes: jokes };
}

export const addJokes = (jokes: Joke[]): JokesActionAdd => {
  return { type: ADD_JOKES, jokes: jokes };
}

export const fetchJokes = (
  jokesLastID: number,
  jokesAmount: number
): JokesActionFetch => {
  return {
    type: FETCH_JOKES,
    props: {
      jokesLastID: jokesLastID,
      jokesAmount: jokesAmount,
    }
  };
}

export function* fetchJokesWorker(action: JokesActionFetch): SagaIterator {
  const data: Joke[] = yield call(
    fetchJokesFromApi,
    action.props.jokesLastID,
    action.props.jokesAmount
  );
  yield put(addJokes(data));
}

export function* jokesWatcher() {
  yield takeEvery(FETCH_JOKES, fetchJokesWorker);
}