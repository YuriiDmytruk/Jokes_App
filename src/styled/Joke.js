import styled from 'styled-components';

export const CardSize = styled.div`
  width: 17rem;
  height: 10rem;
  margin: 10px auto;
`;

export const CardTextSize = styled.span`
  height: 3rem;
`;

export const CenterButton = styled.div`
  width: 100%;
  height: 100%;
  display: flex;
  align-items: center;
  justify-content: center;
`;
