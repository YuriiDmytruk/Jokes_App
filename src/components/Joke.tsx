import * as React from 'react';
import { Link } from 'react-router-dom';
import { Button, Card, Col, Container, Row } from 'react-bootstrap';
import { CardSize, CardTextSize, CenterButton } from '../styled/Joke';

import { useDispatch } from 'react-redux';
import { deleteJoke } from '../redux/ducks/jokes';

import { Joke as JokeType } from '../types';

type JokeProps = {
  joke: JokeType;
};

const CHARS_IN_JOKE = 50;
const DOTS = '...';

const Joke = (props: JokeProps): JSX.Element => {
  const dispatch = useDispatch();

  const onDelete = () => {
    dispatch(deleteJoke(props.joke.id));
  };

  return (
    <Col>
      <Container>
        <CardSize>
          <Card>
            <Card.Body>
              <Card.Title>{props.joke.category}</Card.Title>
              <Card.Text>
                <CardTextSize>
                  {props.joke.joke.slice(0, CHARS_IN_JOKE) + DOTS}
                </CardTextSize>
              </Card.Text>
              <Row>
                <Col>
                  <CenterButton>
                    <Link to={`/joke/${props.joke.id}`}>
                      <Button variant="primary">Check</Button>
                    </Link>
                  </CenterButton>
                </Col>
                <Col>
                  <CenterButton>
                    <Button variant="danger" onClick={onDelete}>
                      Delete
                    </Button>
                  </CenterButton>
                </Col>
              </Row>
            </Card.Body>
          </Card>
        </CardSize>
      </Container>
    </Col>
  );
};

export default Joke;
