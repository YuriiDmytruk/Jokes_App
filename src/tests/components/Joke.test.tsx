import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { createStore } from 'redux';
import { MemoryRouter } from 'react-router-dom';

import Joke from '../../components/Joke';

import { jokesReducer } from '../../redux/ducks/jokes';
import { Joke as JokeType } from '../../types';

test('should render Joke component and match snapshot', () => {
  const store = createStore(jokesReducer, { jokes: [] });

  const joke: JokeType = {
    joke: 'SomeJoke',
    category: 'category',
    id: 0,
  };

  const { asFragment } = render(
    <Provider store={store}>
      <MemoryRouter>
        <Joke joke={joke} />
      </MemoryRouter>
    </Provider>
  );

  expect(asFragment()).toMatchSnapshot();
});
