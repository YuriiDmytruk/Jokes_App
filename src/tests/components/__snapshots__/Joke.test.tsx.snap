// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Joke correctly 1`] = `
<DocumentFragment>
  <div
    class="col"
  >
    <div
      class="container"
    >
      <div
        class="sc-beyTiQ idbsSJ"
      >
        <div
          class="card"
        >
          <div
            class="card-body"
          >
            <div
              class="card-title h5"
            >
              category
            </div>
            <p
              class="card-text"
            >
              <span
                class="sc-guDMob kebUtl"
              >
                SomeJoke...
              </span>
            </p>
            <div
              class="row"
            >
              <div
                class="col"
              >
                <div
                  class="sc-dmyDmy gPGhxd"
                >
                  <a
                    href="/joke/0"
                  >
                    <button
                      class="btn btn-primary"
                      type="button"
                    >
                      Check
                    </button>
                  </a>
                </div>
              </div>
              <div
                class="col"
              >
                <div
                  class="sc-dmyDmy gPGhxd"
                >
                  <button
                    class="btn btn-danger"
                    type="button"
                  >
                    Delete
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
